### ISSUE TITLE
华为AI 证刷题

### ISSUE THREAD

**Issue by 467480903** [467480903/blog#4]
26号晚上考试， 只有2天时间刷题了， 靠北啊

**Comment by 467480903** [comment]
# 聚类任务

聚类任务（Clustering） 是机器学习中最重要的一种无监督学习任务。
它的核心目标是：在没有任何标签（label）的情况下，根据数据点之间的相似性，自动把数据分成若干个组（称为“簇”或“cluster”），使得：

同一个簇内的数据点尽可能相似
不同簇之间的数据点尽可能不同

<img width="1010" height="455" alt="Image" src="https://github.com/user-attachments/assets/948842bb-1844-4cf2-8c81-2337171425c1" />

## 原型聚类， 层次聚类， 密度聚类


### 层次聚类

<img width="846" height="710" alt="Image" src="https://github.com/user-attachments/assets/943586a9-b6b2-42a2-b777-145b2f14f77c" />


#### 凝聚式（Agglomerative）  和   分裂式（Divisive） 
```
sklearn / scipy ('complete')
sklearn / scipy ('average')
sklearn / scipy ('single')
```



### 密度聚类

<img width="870" height="430" alt="Image" src="https://github.com/user-attachments/assets/29878637-ce3c-40d2-92a0-693653cbb96f" />


## K-means 算法

<img width="458" height="181" alt="Image" src="https://github.com/user-attachments/assets/5d506f1b-68aa-4011-9a9f-408df4bbab94" />

```
输入：样本集
过程：
第一步随机选择K个点作为初始原型；
第二步计算剩余的点到K个点的距离，并将点划入距离最近的原型所在的“簇”内；
第三步计算每个“簇”点的均值，并将每个“簇”的均值更新为新的原型；
第四步，重复前面3步，直到每个“簇”点的均值不再变化；
第五步，输出每个"簇"的中心。
为避免运行时间过长，通常会设置一个最大的迭代轮次或者设置最小调整幅度阀值
```

### Voronoi diagram 点集空间分割， 沃罗诺伊图 ， 泰森多边形

<img width="305" height="365" alt="Image" src="https://github.com/user-attachments/assets/3b534ec2-86a7-4407-8843-2a5d4d902719" />

#### Delaunay triangulation ， 点集三角化

<img width="346" height="399" alt="Image" src="https://github.com/user-attachments/assets/b07dec93-4b35-43e5-8f26-2873cccd6206" />

### Vector quantization  矢量量化

数据压缩分类

## Webor problem , 韦伯问题， 选址问题，


### Fermat problem ，费马问题， 3角顶点最短距离求解

<img width="383" height="936" alt="Image" src="https://github.com/user-attachments/assets/8329283d-79c1-42d8-b6fd-40f75051c6e9" />

### attraction-repulsion triangle problem， 吸引-排斥 问题

<img width="510" height="655" alt="Image" src="https://github.com/user-attachments/assets/de60e4ce-5292-4f85-a190-b4749062e858" />

**Comment by 467480903** [comment]
# DBSCAN（Density-Based Spatial Clustering of Applications with Noise）  ， 密度聚类算法


<img width="555" height="599" alt="Image" src="https://github.com/user-attachments/assets/5091bb4f-19d4-4a9b-a0c6-98cba7dc235d" />

<img width="1014" height="701" alt="Image" src="https://github.com/user-attachments/assets/1aa97f5d-5e3d-4cb7-9236-1503a7c5c3e5" />

## 核心点， 边界点， 噪声点

## 密度直达，密度可达， 密度相连， 非密度相连

**Comment by 467480903** [comment]
# DNN 算法


## RNN recurrent neural network 循环神经网络

  具备 记忆功能
  
  由  输入层， 输出层， 循环连接  组成

<img width="1036" height="713" alt="Image" src="https://github.com/user-attachments/assets/0103e139-5051-46d3-8976-fc44c3a1c05e" />

**Comment by 467480903** [comment]
# PCA ， Principal Component Analysis,  主成分分析

数据降维方法

<img width="425" height="576" alt="Image" src="https://github.com/user-attachments/assets/134232da-0f58-4aeb-851d-84ac2f9d4cb6" />

**Comment by 467480903** [comment]
# LSTM 网络 ， long short term memory , 长短期记忆网络， RNN 一种实现

主要由 输入，输出，遗忘  三层 组成，

<img width="670" height="364" alt="Image" src="https://github.com/user-attachments/assets/9fc06e83-2455-48ff-93cc-d272f1037216" />

<img width="1294" height="924" alt="Image" src="https://github.com/user-attachments/assets/ce6042fc-cd2d-41fe-b1a0-287513da4cc0" />

**Comment by 467480903** [comment]
# 梯度累积

梯度累积（Gradient Accumulation） 不是一个独立的“算法”，而是一种非常实用的训练技巧，专门用来解决“显存不够用”的问题，让你能在显存有限的条件下训练更大的模型或更大的 batch size。

一句话总结它的核心作用：

把原本需要一次大 batch 的梯度计算，拆成多次小 batch 累加，再一次性更新参数，达到“模拟大 batch size”的效果。

**Comment by 467480903** [comment]
# ASIC  和 FPGA
Application-Specific Integrated Circuit，专用集成电路,
某个特定应用量身定制的芯片，一次性烧录，出厂后不可更改 

Field-Programmable Gate Array，现场可编程门阵列
出厂时是“白板”，用户可以现场多次编程配置，像“乐高积木”或“可重编程的万能板”。

**Comment by 467480903** [comment]
# 华为AI F1

<img width="1108" height="608" alt="Image" src="https://github.com/user-attachments/assets/bd9bb10d-eb3a-47be-b2b7-2ec167febd6f" />

**Comment by 467480903** [comment]
# 决策树 ， desision tree

监督类学习的一种方式， 有 CART , ID, C 
Classification And Regression Tree ，分类与回归树，  二叉树类， 
Iterative Dichotomiser 3， 叠代二分器， 
OC1 (Oblique classifier 1)
[Chi-square automatic interaction detection](https://en.wikipedia.org/wiki/Chi-square_automatic_interaction_detection) (CHAID)

**Comment by 467480903** [comment]
# 模型误差

泛化误差 = 方差^2+偏差+不可约误差

**Comment by 467480903** [comment]
# CNN, 卷积神经网络， Convolutional Neural Network

一种前馈神经网络，  是监督类学习AI

CNN = 输入层 + 隐藏层+输出层

隐藏层=卷积层， 池化层， 全连接层 ， 归一化层， 激活函数

被广泛用于图像识别， 典型应用在yolo 上， 

## 卷积层（Convolutional Layer）
从输入数据中提取局部特征

在输入特征图上滑动，以固定大小的窗口执行加权和运算

<img width="445" height="86" alt="Image" src="https://github.com/user-attachments/assets/7704021a-1ed3-4e6d-8459-0b7ec3b5ac14" />

w是卷积核的权重，b 为偏置项，x 是输入特征图，y 是输出特征图

### 特殊卷积类型
单位卷积（1×1 卷积）：用于减少通道数或构建网络中的全连接结构，常用于 Network-In-Network（NIN）或 MobileNet 中。

平铺卷积（tiled convolution）：卷积核仅在图像部分区域滑动，参数部分共享，用于捕捉图像中的平移不变性。

反卷积（deconvolution）/转置卷积（transposed convolution）：用于上采样操作，常见于语义分割与生成模型。

扩张卷积（dilated convolution）：在卷积核内部插入空隙扩大感受野，用于建模长程依赖，特别在自然语言处理（NLP）和语音识别中表现突出。

### 卷积层参数
卷积层的核心超参数包括卷积核大小、步长（stride）与填充（padding），它们共同决定输出特征图的尺寸。

卷积核大小 f：决定每次操作的感受野范围，卷积核越大可提取更全局的特征。

步长 s：表示卷积核滑动的步幅，步长越大输出尺寸越小。

填充 p：为保持特征图尺寸或避免边缘信息损失，在输入边缘填充零或复制边界值

## 激活函数  ReLU（Rectified Linear Unit）

卷积层后的非线性变换通常由激活函数完成，用于增强模型的表达能力

<img width="285" height="65" alt="Image" src="https://github.com/user-attachments/assets/f4e6f376-e154-4b77-9f34-7738a3cd4091" />

## 池化层 Pooling Layer

池化层用于对特征图进行下采样，减少参数数量，提升特征不变性

LP 池化

<img width="181" height="84" alt="Image" src="https://github.com/user-attachments/assets/620c3401-b961-4ae6-a0df-166feb7269f7" />

**Comment by 467480903** [comment]
# 梯度消失  梯度爆炸

梯度消失（Vanishing Gradient）和梯度爆炸（Exploding Gradient） 是训练深层网络时最经典、最常见的两大优化难题

<img width="993" height="290" alt="Image" src="https://github.com/user-attachments/assets/21d9fc12-cf2c-40ac-ac34-aae55ade46ed" />

# 全局梯度下降算法、随机梯度下降法和批量梯度下降算法

<img width="1003" height="226" alt="Image" src="https://github.com/user-attachments/assets/551ab43a-d207-41f5-ba2f-d867d175e1e3" />

批量， Batch Gradient Descent
随机，Stochastic Gradient Descent, SGD
小批量， Mini-batch Gradient Descent

**Comment by 467480903** [comment]
#   联算比失衡

也叫计算-通信比失衡、Compute-Communication Imbalance，或更通俗的说法“算力吃饱了，通信卡住了

现在的 GPU 算力提供商， 会有几千上万张GPU卡， 集中运维， 租赁给外部客户，

Oracle 有 13万张 BlackWell ， 马斯克xAI有 10万张， TenserWave 有 8000张AM卡， 

这样租赁显卡使用的时候， 发现超大模型训练或微调时， 会发生  服务器间网速影响显卡使用性能  这个现象

**Comment by 467480903** [comment]
#  CNN 中的池化技术

池化（Pooling）是卷积神经网络（CNN）中非常重要的操作，主要用于降采样、减少计算量、增大感受野、提供平移不变性等

## 全局池化

Global pooling， 提高计算性能 ， 会丢失空间信息， 

## 最大池化

max pooling 选择特征最明显的区域处理， 会丢失细节

## 平均池化
min pooling , 平滑特征， 避免细节丢失


## 随机池化
Stochastic Pooling ， 现在很少用

**Comment by 467480903** [comment]
#  模型过拟合 ，Overfitting

模型在训练数据上表现极好（训练误差很低），但在未见过的新数据（测试集/真实世界数据）上表现很差（泛化误差很高）。

<img width="984" height="209" alt="Image" src="https://github.com/user-attachments/assets/bbbc58a6-5c86-4a26-8a53-8482a7361419" />

<img width="1031" height="689" alt="Image" src="https://github.com/user-attachments/assets/7a418108-e4a4-456e-9ac1-7b6f58a8a9f0" />

**Comment by 467480903** [comment]
# 模型蒸馏

用一个大模型（Teacher，老师模型）来“教”一个小模型（Student，学生模型），让小模型学习大模型的知识，从而在参数量小、推理速度快的情况下，尽可能接近大模型的性能

<img width="854" height="740" alt="Image" src="https://github.com/user-attachments/assets/cca8e07b-f059-48bb-953d-0a26b8f8dea2" />

**Comment by 467480903** [comment]
# 动态计算图

动态计算图（Dynamic Computation Graph）和静态计算图（Static Computation Graph）是深度学习框架中两种最主要的计算图构建方式，它们的主要区别在于计算图是在什么时候构建的，以及构建后是否可以修改。

新的框架都支持 动态计算图
ONNX 不支持动态计算， 只能静态， 
华为的 MindSpore也只能静态

**Comment by 467480903** [comment]
# seq2seq 

Sequence-to-Sequence，序列到序列）是深度学习中最经典、最重要的模型架构之一，主要用于处理输入和输出都是变长序列的任务

**Comment by 467480903** [comment]
<img width="802" height="484" alt="Image" src="https://github.com/user-attachments/assets/d597c38f-6091-41ef-b72e-474df143f386" />
